server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/data/StarTreeDataWriter.java

```diff
@@ -66,21 +66,32 @@ private static void writeStarTreeNodes(IndexOutput output, InMemoryTreeNode root
         while (!queue.isEmpty()) {
             InMemoryTreeNode node = queue.remove();
 
-            if (node.children == null || node.children.isEmpty()) {
+            if ((node.children == null || node.children.isEmpty()) && node.childStarNode == null) {
```

Comment:
Can we add a method in node like `hasChild()` which can combine all these? We should not leak the internal arrangement within the class to the writer